src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+            raise ValueError("resource count must be > 0")
+
+        res = {"type": res_type, "count": count}
+
+        if with_child:
+            res["with"] = with_child
+        return res
+
+    @classmethod
+    def _create_slot(cls, label, count, with_child):
+        slot = cls._create_resource("slot", count, with_child)
+        slot["label"] = label
+        return slot
+
+    @property
+    def duration(self):
+        try:
+            return self.jobspec["attributes"]["system"]["duration"]
+        except KeyError:
+            return None
+
+    @duration.setter
+    def duration(self, duration):
+        """
+        Assign a time limit to the job.  The duration may be:
+        - a float in seconds
+        - a string in Flux Standard Duration
+        A duration of zero is interpreted as "not set".
+        """
+        if isinstance(duration, six.string_types):
+            time = parse_fsd(duration)
+        elif isinstance(duration, float):
+            time = duration
+        else:
+            raise ValueError("duration must be a float or string")
+        if time < 0:
+            raise ValueError("duration must not be negative")
+        if math.isnan(time) or math.isinf(time):
+            raise ValueError("duration must be a normal, finite value")
+        self.jobspec["attributes"]["system"]["duration"] = time
+
+    @property
+    def cwd(self):
+        """
+        Get working directory of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["cwd"]
+        except KeyError:
+            return
+
+    @cwd.setter
+    def cwd(self, cwd):
+        """
+        Set working directory of job.
+        """
+        if not isinstance(cwd, six.string_types):
+            raise ValueError("cwd must be a string")
+        self.jobspec["attributes"]["system"]["cwd"] = cwd
+
+    @property
+    def environment(self):
+        """
+        Get (entire) environment of job.
+        """
+        try:
+            self.jobspec["attributes"]["system"]["environment"]
+        except KeyError:
+            return
+
+    @environment.setter
+    def environment(self, environ):
+        """
+        Set (entire) environment of job.
+        """
+        if not isinstance(environ, collectionsAbc.Mapping):
+            raise ValueError("environment must be a mapping")
+        self.jobspec["attributes"]["system"]["environment"] = environ
+
+    def _set_treedict(self, in_dict, key, val):
+        """
+        _set_treedict(d, "a.b.c", 42) is like d[a][b][c] = 42
+        but levels are created on demand.
+        """
+        path = key.split(".", 1)
+        if len(path) == 2:
+            self._set_treedict(in_dict.setdefault(path[0], {}), path[1], val)
+        else:
+            in_dict[key] = val
+
+    def setattr(self, key, val):
+        """
+        set job attribute
+        """
+        self._set_treedict(self.jobspec, "attributes." + key, val)
+
+    def setattr_shopt(self, key, val):
```

Comment:
Perhaps personal preference, but `setattr_shell_option` would feel better to me here.  The `shopt` suffix isn't immediately obvious to me waht it is.